### ISSUE TITLE
Out of Memory Bug

### ISSUE THREAD

**Issue by jiruijin** [usnistgov/alignn#93]
I trained the model with 28000 cif data, but every time I ran it, I got an error:
 "slurmstepd: error: Detected 1 oom-kill event(s) in StepId=59934605.batch. Some of your processes may have been killed by the cgroup out-of-memory handler. "
I already used 500GB for CPU, why is it still out of memory?
![image](https://user-images.githubusercontent.com/112404714/226739093-0a4a01a8-5b34-4d6c-96fa-46318d447bcc.png)

**Comment by jiruijin** [comment]
Thanks a lot for answering so fast @bdecost, but I think when you trained the model for the material project dataset, it should be a larger dataset than me. And I tried with a small dataset like 4000 cif files, it worked. It seems that currently, I can only decrease the size of dataset, hope you can fix it soon.

**Comment by jiruijin** [comment]
The dataset I am using is from oqmd, half of them (12000) only have 5 five atoms. In the rest of the dataset, the range for atoms number is from 10 to 40.

**Comment by knc6** [comment]
You can lower the batch size to something like 2 or 5 , which should resolve the issue.

**Comment by bdecost** [comment]
can you clarify exactly when the resource manager is killing your job? from your screen cap it looks like it's during the dataloader setup before any data is sent to the GPU?

> The dataset I am using is from oqmd, half of them (12000) only have 5 five atoms. In the rest of the dataset, the range for atoms number is from 10 to 40.

I'm a little surprised this dataset is hitting your memory limit, but I will investigate. Do you mind sharing the slurm configuration you're using and maybe the limits of the partition you're running on?

**Comment by bdecost** [comment]
to clarify a bit, if GPU memory limit is the issue, you'll get a CUDA out of memory error, and slurm will kill your job because the training script crashed. I think you are running into a memory vs compute tradeoff that we made because our graph construction code was slow. I'm working on fixing that, but there are a couple other fixes I'm trying to do at the same time...

**Comment by jiruijin** [comment]
![image](https://user-images.githubusercontent.com/112404714/226947733-7d9d909e-1de9-42f6-b2c4-f69483f73b83.png)
Thank you for the information. Above is the slurm configuration I am using. And the problem is: 
![image](https://user-images.githubusercontent.com/112404714/226949196-27f7aaa9-2973-43c7-961d-9a3c7dc1783a.png)

**Comment by bdecost** [comment]
right, ok. How many dataloader workers are you using? Each worker process [apparently makes a full copy](https://discuss.pytorch.org/t/is-a-dataset-copied-as-part-of-dataloader-with-multiple-workers/112924/2) of the dataset, so if you're using multiple you can try to reduce that for now as a band-aid

That's not a real solution, but I'll try to get a fix out this week or next. There are a few things to do, most importantly including computing the graphs during minibatch construction instead of caching them all in memory

**Comment by jiruijin** [comment]
I am using the default setting: num_workers: int = 4. I will decrease it and retry. Thank you!

**Comment by jiruijin** [comment]
I used get_primitive_structure function in Pymatgen for all my cif files and set num_workers = 10 in the config file, then the problem is solved!
